query.go

```diff
@@ -0,0 +1,150 @@
+package papergres
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/jmoiron/sqlx"
+)
+
+// Query holds the SQL to execute and the connection string
+type Query struct {
+	SQL      string
+	Database *Database
+	Args     []interface{}
+	insert   bool
+}
+
+// SelectParamsFn is a function that takes in the iteration and
+// returns the destination and args for a SQL execution.
+type SelectParamsFn func(i int) (dest interface{}, args []interface{})
+
+// Exec runs a sql command given a connection and expects LastInsertId or RowsAffected
+// to be returned by the script. Use this for INSERTs
+func (q *Query) Exec() *Result {
+	return exec(q, false)
+}
+
+// ExecAll gets many rows and populates the given slice
+// dest should be a pointer to a slice
+func (q *Query) ExecAll(dest interface{}) *Result {
+	all := func(db *sqlx.DB, r *Result) error {
+		err := db.Select(dest, q.SQL, q.Args...)
+
+		r.RowsReturned = getLen(dest)
+
+		return err
+	}
+
+	return execDB(q, all)
+}
+
+// ExecAllIn works with IN queries. It will take a slice of values and
+// attach the slice to the query as a list of values.
+// One key difference with bindvar used for IN query is a `?` (question mark)
+// the query then has to be rebinded to change default bindvar to target bindvar
+// like `$1` (dollar sign followed by a number) for postgres etc.
+func (q *Query) ExecAllIn(dest interface{}) *Result {
+	all := func(db *sqlx.DB, r *Result) error {
+		query, args, err := sqlx.In(q.SQL, q.Args...)
+		if err != nil {
+			return err
+		}
+
+		// Rebind the query to replace the ? with $1, $2, etc
+		query = db.Rebind(query)
+
+		// Execute a select query using this DB
+		err = db.Select(dest, query, args...)
+
+		r.RowsReturned = getLen(dest)
+
+		return err
+	}
+
+	return execDB(q, all)
+}
+
+// ExecNonQuery runs the SQL and doesn't look for any results
+func (q *Query) ExecNonQuery() *Result {
+	return exec(q, true)
+}
+
+// ExecSingle fetches a single row from the database and puts it into dest.
+// If more than 1 row is returned it takes the first one.
+// Expects at least 1 row or it will error.
```

Comment:
```Golang
// Expects at least 1 row or it will error.
```
to
```Golang
// Expects at least 1 row or it will return an error.
```

Reply:
:+1:  Nice. Updated the test in a new commit.